packages/server/lib/crons/deleteOldData.ts

```diff
@@ -50,62 +70,226 @@ export async function exec(): Promise<void> {
+
+                    // Delete connections
+                    await batchDelete({
+                        name: 'connections < integration',
+                        startTimestamp,
+                        ttlMs,
+                        deleteFn: async () => await connectionService.hardDeleteByIntegration({ limit, integrationId: integration.id! })
+                    });
+
+                    await configService.hardDelete(integration.id!);
+                }
+
+                return integrations.length;
+            }
+        });
+
+        // Delete sync configs and all associated data
+        await batchDelete({
+            name: 'sync configs',
+            startTimestamp,
+            ttlMs,
+            deleteFn: async () => {
+                const syncsConfigs = await getSoftDeletedSyncConfig({ limit, olderThan: deleteSyncConfigsOlderThan });
+
+                for (const syncConfig of syncsConfigs) {
+                    await deleteSyncConfigData({ syncConfig, startTimestamp, ttlMs });
+                }
+
+                return syncsConfigs.length;
             }
-            if (Date.now() - startTimestamp > ttlMs) {
-                logger.info(`Time limit reached, stopping`);
-                return;
+        });
+
+        // Delete connections and all associated data
+        await batchDelete({
+            name: 'connections',
+            startTimestamp,
+            ttlMs,
+            deleteFn: async () => {
+                const connections = await connectionService.getSoftDeleted({ limit, olderThan: deleteConnectionsOlderThan });
+
+                for (const connection of connections) {
+                    logger.info('Deleting connection...', connection.id, connection.connection_id);
+                    const resSyncs = await db.knex
+                        .select<
+                            {
+                                sync: Sync;
+                                syncConfig: DBSyncConfig;
+                            }[]
+                        >(db.knex.raw('row_to_json(_nango_syncs.*) as sync'), db.knex.raw('row_to_json(_nango_sync_configs.*) as "syncConfig"'))
+                        .from<Sync>('_nango_syncs')
+                        .join('_nango_sync_configs', '_nango_sync_configs.id', '_nango_syncs.sync_config_id')
+                        .where({ nango_connection_id: connection.id });
+                    for (const res of resSyncs) {
+                        await deleteSyncData({ ...res });
+                    }
+
+                    await connectionService.hardDelete(connection.id);
+                }
+
+                return connections.length;
             }
-            await setTimeout(1000);
-        }
+        });
     } finally {
         if (lock) {
             locking.release(lock);
         }
     }
 }
+
+async function batchDelete({
+    name,
+    deleteFn,
+    startTimestamp,
+    ttlMs
+}: {
+    name: string;
+    deleteFn: () => Promise<number>;
```

Comment:
we have to be careful when passing deleteFn: if an entity is used to retrieve children, we must ensure that the entity is deleted last. Otherwise if the deletion is interrupted by a exception or expires we wouldn't be able to retrieve the childer again and end up with orphans.

Reply:
this is what happened yes, we always delete at the end (unless mistaken)